pkg/publicapi/middleware/logger.go

```diff
@@ -0,0 +1,85 @@
+package middleware
+
+import (
+	"net/http"
+	"time"
+
+	"github.com/bacalhau-project/bacalhau/pkg/publicapi/apimodels"
+	chimiddleware "github.com/go-chi/chi/v5/middleware"
+	"github.com/rs/zerolog"
+	"github.com/rs/zerolog/log"
+)
+
+type ZeroLogFormatter struct {
+	logger            *zerolog.Logger
+	onlyErrorStatuses bool
+}
+
+type ZeroLogFormatterOption func(*ZeroLogFormatter)
+
+// WithLogger sets a logger for the ZeroLogFormatter.
+func WithLogger(logger *zerolog.Logger) ZeroLogFormatterOption {
+	return func(z *ZeroLogFormatter) {
+		z.logger = logger
+	}
+}
+
+// WithOnlyErrorStatuses sets the onlyErrorStatuses flag for the ZeroLogFormatter.
+func WithOnlyErrorStatuses(errorsOnly bool) ZeroLogFormatterOption {
+	return func(z *ZeroLogFormatter) {
+		z.onlyErrorStatuses = errorsOnly
+	}
+}
+
+// NewZeroLogFormatter returns a new ZeroLogFormatter configured with the provided option setters.
+func NewZeroLogFormatter(options ...ZeroLogFormatterOption) *ZeroLogFormatter {
+	// default values
+	formatter := &ZeroLogFormatter{
+		logger:            &log.Logger,
+		onlyErrorStatuses: true,
+	}
+
+	// apply the options
+	for _, option := range options {
+		option(formatter)
+	}
+
+	return formatter
+}
+
+// NewLogEntry returns a new LogEntry for the request.
+func (l *ZeroLogFormatter) NewLogEntry(r *http.Request) chimiddleware.LogEntry {
+	return zeroLogEntry{
+		request:   r,
+		formatter: l,
+	}
+}
+
+type zeroLogEntry struct {
+	request   *http.Request
+	formatter *ZeroLogFormatter
+}
+
+// Write implements the io.Writer interface to write a string to the logger.
+func (l zeroLogEntry) Write(status, bytes int, header http.Header, elapsed time.Duration, extra interface{}) {
+	if l.formatter.onlyErrorStatuses && status < http.StatusBadRequest {
+		return
+	}
+	l.formatter.logger.Info().
```

Comment:
The idea was to only log responses with error codes (>=400) in dev and test stacks, and ignore successful requests. I have improved things in [9a9f3c6](https://github.com/bacalhau-project/bacalhau/pull/2772/commits/9a9f3c6a9cd65cd4b8e1e159ca36f291659daebb) by moving this weird logic into `node` package, and allow the logger to define the level it should log at. By default, it will log 2xx at info, 4xx at warn and 5xx at error.